apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2466,25 +2484,102 @@ func (u *UniterAPI) GoalStates(args params.Entities) (string, error) {
 			result.Results[i].Error = common.ServerError(common.ErrPerm)
 			continue
 		}
-		// TODO (agprado): get units and relations information from
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		application, err := unit.Application()
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		} else {
+			result.Results[i].Result, err = u.goalStateResult(application)
+			if err != nil {
+				result.Results[i].Error = common.ServerError(err)
+			}
+		}
+	}
+	return result, nil
+}
 
-		result.Results[0].Result = "Hello World I'll be a yaml"
+// goalStateResult creates the structure with an application units and unit relations.
+func (u *UniterAPI) goalStateResult(application *state.Application) (*params.GoalState, error) {
+	gs := params.GoalState{}
+
+	var err error
+	gs.Units, err = u.goalStateUnits(application)
+	if err != nil {
+		return nil, err
 	}
-	return result.Results[0].Result, nil
+	allRelations, err := application.Relations()
+	if err != nil {
+		return nil, err
+	}
+	if allRelations != nil {
+		gs.Relations, err = u.goalStateRelations(allRelations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &gs, nil
 }
 
-// GetCloudSpec returns the cloud spec used by the model in which the
-// authenticated unit or application resides.
-// A check is made beforehand to ensure that the request is made by an entity
-// that has been granted the appropriate trust.
-func (u *UniterAPI) CloudSpec() (params.CloudSpecResult, error) {
-	canAccess, err := u.accessCloudSpec()
+// goalStateRelations creates the structure with all the relations between endpoints in an application.
+func (u *UniterAPI) goalStateRelations(allRelations []*state.Relation) (map[string]params.UnitsGoalState, error) {
+
+	result := map[string]params.UnitsGoalState{}
+
+	for _, r := range allRelations {
+
+		endPoints := r.Endpoints()
+		for _, e := range endPoints {
+			if e.Relation.Role == "peer" {
+				continue
+			}
+			application, err := u.st.Application(e.ApplicationName)
+			if err != nil {
+				return nil, err
+			}
+			units, err := u.goalStateUnits(application)
+			if err != nil {
+				return nil, err
+			}
+			if len(units) == 0 {
+				continue
+			}
+			if result[e.Name] == nil {
+				result[e.Name] = params.UnitsGoalState{}
+			}
+			relation := result[e.Name]
+			for unitName, unitGS := range units {
+				relation[unitName] = unitGS
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// goalStateUnits loops through all application units, and stores the
+// application GoalStateStatus in UnitsGoalState.
+func (u *UniterAPI) goalStateUnits(application *state.Application) (params.UnitsGoalState, error) {
+
+	allUnits, err := application.AllUnits()
 	if err != nil {
-		return params.CloudSpecResult{}, err
+		return nil, err
 	}
-	if !canAccess() {
-		return params.CloudSpecResult{Error: common.ServerError(common.ErrPerm)}, nil
+
+	units := params.UnitsGoalState{}
```

Comment:
Probably better to call this unitsGoalState otherwise people will think the variable contains units